client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-how-to-migrate/use-pending-migration-status.ts

```diff
@@ -0,0 +1,45 @@
+import { useEffect } from 'react';
+import { useUpdateMigrationStatus } from 'calypso/data/site-migration/use-update-migration-status';
+import { HOW_TO_MIGRATE_OPTIONS } from 'calypso/landing/stepper/constants';
+import { useSite } from 'calypso/landing/stepper/hooks/use-site';
+import type { NavigationControls } from '../../types';
+
+interface PendingMigrationStatusProps {
+	onSubmit?: Pick< NavigationControls, 'submit' >[ 'submit' ];
+}
+
+const usePendingMigrationStatus = ( { onSubmit }: PendingMigrationStatusProps ) => {
+	const site = useSite();
+	const siteId = site?.ID;
+
+	const canInstallPlugins = site?.plan?.features?.active.find(
+		( feature ) => feature === 'install-plugins'
+	)
+		? true
+		: false;
+
+	const { updateMigrationStatus } = useUpdateMigrationStatus();
+
+	// Register pending migration status when loading the step.
+	useEffect( () => {
+		if ( siteId ) {
+			updateMigrationStatus( siteId, 'migration-pending' );
+		}
+	}, [ siteId, updateMigrationStatus ] );
+
+	const setPendingMigration = ( how: string ) => {
+		const destination = canInstallPlugins ? 'migrate' : 'upgrade';
+		if ( siteId ) {
+			const parsedHow = how === HOW_TO_MIGRATE_OPTIONS.DO_IT_MYSELF ? 'diy' : how;
+			updateMigrationStatus( siteId, `migration-pending-${ parsedHow }` );
+		}
+
+		if ( onSubmit ) {
+			return onSubmit( { how, destination } );
+		}
+	};
+
+	return { setPendingMigration };
+};
```

Comment:
Hey folks! I stumbled upon this code today; I got really confused. I don't understand why we'd wrap the existing `useUpdateMigrationStatus` in another hook. Can you enlighten me on the reasoning for this? Even for me, who worked on this, I had difficulties finding where the status was being updated.

Especially since the hook is strictly tightened with the How to migrate step, as it won't work anywhere else.

I also disagree that the action on the `How to migrate` step is to "set a migration as pending", and that's the idea that it we give when we do:
```
onClick={ () => setPendingMigration( option.value ) }
```
For me, setting the pending state is a side effect of the submit action and not the other way around. I also feel that for future maintenance, removing the submit logic from the How to migrate step file just makes it harder to understand the code. We added a new layer for the `onSubmit` event, which by itself is already deep enough.

Reply:
I agree that it seems not necessary  to have the `submit inside the `usePendingMigrationStatus` 

> Hey folks! I stumbled upon this code today; I got really confused. I don't understand why we'd wrap the existing useUpdateMigrationStatus in another hook.

Another point is that we are not waiting for the request to be finished before calling the onSubmit. 


> I don't understand why we'd wrap the existing useUpdateMigrationStatus in another hook.
I am ok with the extraction because it removes business logic from the original component and prevents some components from re-rendering. For example, useEffects inside hooks don't trigger component rendering. 

But I am unaware of why we need to use 2 states to set migration as pending.

Reply:
> But I am unaware of why we need to use 2 states to set migration as pending.

I think the first one is to flag when the user gets to the How to migrate step so we can redirect them back to that stage. The `migration-pending-{difm/diy}` means that the user has already submitted the intent.

Reply:
> For example, useEffects inside hooks don't trigger component rendering.

In this case, would it trigger the render? I asked this because we are not setting any states that would justify that.

Reply:
> I don't understand why we'd wrap the existing useUpdateMigrationStatus in another hook

The idea was to extract the portion of logic related to the sticker creation in the how to migrate. Otherwise, it would introduce a big complexity to the `SiteMigrationHowToMigrate`, which already has many other logic too. I thought having it in the same folder as the component would be enough, but maybe we could try to rename the hook to something else to avoid the confusion? I tried to think something to add the HowToMigrate in the name, but my ideas seem too verbose. Something like `usePendingMigrationStatusHandlerForHowToMigrateStep`?

> I also disagree that the action on the How to migrate step is to "set a migration as pending", and that's the idea that it we give when we do:
> 
> `onClick={ () => setPendingMigration( option.value ) }`
> 
> For me, setting the pending state is a side effect of the submit action and not the other way around. I also feel that for future maintenance, removing the submit logic from the How to migrate step file just makes it harder to understand the code. We added a new layer for the onSubmit event, which by itself is already deep enough.

Hm! Looking at this again, I totally agree with you. Maybe we could refator it removing the `onSubmit` dependency from that and adding a clickHandler that will call the `setPendingMigration` and the `navigation.submit`?

> Another point is that we are not waiting for the request to be finished before calling the onSubmit.

This is already being fixed as part of https://github.com/Automattic/wp-calypso/pull/95612
Also, notice that any refactor would be nice to do on top of this last PR to avoid conflicts. 👆 @m1r0, I think we could merge that, right?

> I think the first one is to flag when the user gets to the How to migrate step so we can redirect them back to that stage. The migration-pending-{difm/diy} means that the user has already submitted the intent.

That's correct!

> In this case, would it trigger the render? I asked this because we are not setting any states that would justify that.

That's less about it and more about the separation of concerns to simplify the component logic and avoid the components with hundreds of lines complex to maintain! 🤭

Reply:
> I think we could merge that, right?

Yep, I've merged it. 👍

Reply:
Also, one important thing when moving the submit to the component, it needs to happen after the sticker is created. 😉

The https://github.com/Automattic/wp-calypso/pull/95612 does it fixing a race condition issue.